### PR TITLE
Fix/authentication/sync-problems : Fix sync problem with auth screens

### PR DESCRIPTION
**Pull request by charliemangano**
# Fix sync problem with auth screens

## Description
This PR introduces a fix for issue #187: it solves the syncing problem that made a user have to click two times on the "SignUp"/"SignIn" button on the authentication screens.

The problem came from how we handle our side effects, as the callbacks of the view model execute, at least partly, only after we've checked in the authentication screens if the loging in/signing up was successful, for example:

```kotlin
// in `SignIn`
authenticationViewModel.logInWithEmail(email, password)            // this does not finish executing

val loginSuccess = userState is UserAuthenticationState.Success    // before this happens
```
So our authentication screens thought that the operation failed, when it had just not finished executing.

## Changes
Changed the signature of the functions in `AuthenticationViewModel` to have `onSuccess` and `onFailure` callbacks. I moved the code that was previously in the button's `onClick` into the callbacks of the calls to `logInWithEmail` and `signUpWithEmail`. This ensures that this bit of code is executed **after** the rest of the `logInWithEmail` or `signUpWithEmail` code, which solves the problem.
This change of signature was applied to every function preventively, and all have simple `Log`s as default values. Docs were updated as well.
Updated the tests for the SignIn and SignUp screens to implement the new callbacks properly.

#### Added
None
#### Modified
- `app/src/main/java/com/android/periodpals/model/authentication/AuthenticationViewModel.kt`: add callbacks to function signatures
- `app/src/main/java/com/android/periodpals/ui/authentication/SignIn.kt`, `app/src/main/java/com/android/periodpals/ui/authentication/SignUp.kt`: updated the calls to the view model functions to implement the callbacks properly
- `app/src/androidTest/java/com/android/periodpals/ui/authentication/SignInTest.kt`, `app/src/androidTest/java/com/android/periodpals/ui/authentication/SignUpTest.kt`: updated the tests to implement the callbacks properly

#### Removed
None

## Dependencies Added
None

## Testing
Existing tests were updated but still check for the same things. Tests all pass. None more written. This should help for the end-to-end tests.

### CODE REVIEW

**Review by francelu — APPROVED** [review]
## Summary
**LGTM !**  
As discussed with the coaches during the sprint meeting, you did the necessary changes to fix this double click issue at the authentication steps !

## Important

#### Code Quality
- New code implemented for this fix is clean, readable and consistent with rest of our code.

#### Functionality
- Code works perfectly, no trouble ! This will be very helpful for our e2e tests I believe !

#### Performance
- No longer the arbitrary thread sleep time !
  *(I admit, it was a very random sleep time that I had implemented on the spot that only fixed half way the double click issue. We hadn't asked the coaches about that issue yet.)*

#### Security
- I don't think applicable. However data is sent to Supabase efficiently !

## Testing
- Tests were already very complete.
- Good adaptation of existing tests for edge cases !

## Documentation
- Good update of existing documentation for modified functions !

## Suggestions
- None.

## Steps before PR merge
- Reply to my general knowledge question about the test and VM :)

File: app/src/androidTest/java/com/android/periodpals/ui/authentication/SignUpTest.kt [hunk 1]
```diff
@@ -47,6 +47,13 @@ class SignUpScreenTest {
     navigationActions = mock(NavigationActions::class.java)
     authViewModel = mock(AuthenticationViewModel::class.java)
 
+    `when`(
+            authViewModel.signUpWithEmail(
+                userEmail = any(), userPassword = any(), onSuccess = any(), onFailure = any()))
+        .thenAnswer {
+          val onSuccess = it.arguments[2] as () -> Unit
+          onSuccess()
+        }
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/androidTest/java/com/android/periodpals/ui/authentication/SignUpTest.kt:56]: We are mocking the view model so that when a view model method/attribute is called, we can see it and control what's returned. The mock doesn't do any of the stuff the original does if not explicited.

The ``when`` on the `authenticationUserState` attribute just returns 'Success`: that means that when we do 
```kotlin
val userState = authenticationViewModel.userAuthenticationState
```
we can be sure that it will set `userState` to `Success`. It doesn't do anything else, and specifically not anything related to the `signUpWithEmail` method.

Now onto the new part of the code: I modified the `onClick` such that the navigation is called inside `signUpWithEmail`'s `onSuccess` callback. That means that when we call `signUpWithEmail`, we want it to execute the `onSuccess` callback, so that the navigation is called properly. The ``when`` statement does just that: when `signUpWithEmail` is called, it gets the argument at index 2 (`it.arguments[2]`), which is the `onSuccess` callback, and then executes it.

Does this make sense?

**Review by coaguila — APPROVED** [review]
lgtm

**Comment by charliemangano** [comment]
Realized a cleanup change that I thought would not affect the behavior actually introduced a bug. Reopening the branch to fix.
